contracts/crowdsale/validation/ControlledAccessCrowdsale.sol

```diff
@@ -0,0 +1,109 @@
+pragma solidity ^0.4.18;
+
+import "../Crowdsale.sol";
+import "../../ownership/Ownable.sol";
+
+
+/**
+ * @title ControlledAccess
+ * @dev The ControlledAccess contract allows functions to be restricted to users
+ * that possess a signed authorization from the owner of the contract. This signed
+ * message includes the address to give permission to and the contract address.
+ * Both addresses are required to prevent reusing the same authorization message
+ * on different contract with same owner.
+ */
+contract ControlledAccessCrowdsale is Crowdsale, Ownable {
+
+  /**
+   * @dev Requires msg.sender to have valid access message.
+   * @param _sig Valid signature from owner 
+   */
+  modifier onlyValidAccess(bytes _sig)
+  {
+    require(isValidAccessMessage(msg.sender, _sig));
+    _;
+  }
+
+  /**
+   * @dev Verifies if message was signed by owner to give access to _add for this contract.
+   *      Assumes Geth signature prefix.
+   * @param _add Address of agent with access.
+   * @param _sig Valid signature from owner 
+   * @return Validity of access message for a given address.
+   */
+  function isValidAccessMessage(address _add, bytes _sig)
+      view public returns (bool)
+  {
+    bytes32 r;
+    bytes32 s;
+    uint8 v;
+
+    //Extract ECDSA signature variables from `sig`
+    assembly {
+      r := mload(add(_sig, 32))
+      s := mload(add(_sig, 64))
+      v := byte(0, mload(add(_sig, 96)))
+    }
+
+    // Version of signature should be 27 or 28, but 0 and 1 are also possible versions
+    if (v < 27) {
+      v += 27;
+    }
+
+    // Verifying if recovered signer is contract owner
+    bytes32 hash = keccak256(this, _add);
+    return owner == ecrecover(
+      keccak256("\x19Ethereum Signed Message:\n32", hash),
+      v,
+      r,
+      s
+    );
+  }
+
+  /**
+   * @dev low level token purchase ***DO NOT OVERRIDE***
+   * @param _beneficiary Address performing the token purchase
+   * @param _sig Access message to provide with purchase.
+   */
+  function buyTokens(address _beneficiary, bytes _sig) public payable {
```

Comment:
giving this contract extends from `Crowdsale` we will still have the inherited `buyTokens` public function which can be called to bypass the signature check, am I right?

Reply:
Notice that there is no need to override this function, in fact there is comment mentioning no to do that 😛 
I think you just need to override the `_preValidatePurchase` function to use your modifier with `msg.data`, am I right?

Reply:
That is correct, could assume signature is msg.data. Will edit to conform with this logic.

Reply:
So, the problem with not overwriting `buyTokens(_beneficiary)` is that you can't call `buyTokens(_beneficiary)` with additionnal msg.data. Data will either be ignored or the function will throw for invalid number of arguments (I believe).

Reply:
Or, in other words, the following is not possible : 

```
await this.crowdsale.buyTokens(buyer,{ from: delegateBuyer, value: amount, data: signature });
```
Where, in this case, msg.data is `0xFUNCTION000000...USERADDRESS` and the signature is omitted.

Reply:
How does solidity handle a data field like

`0x` + `buyTokens function sig` + `_beneficiary addresss` + `signature data`

? will it call `buyTokens` with `_beneficiary` like normal and ignore the `signature data` part? if so, you could, in the pre-validate, pull the signature from `msg.data` and validate from there?